harness/tests/experiment/pytorch/test_deepspeed_trial.py

```diff
@@ -0,0 +1,651 @@
+# type: ignore
+import copy
+import os
+import pathlib
+from typing import Any, Dict, Optional
+
+import pytest
+import torch
+
+import determined
+import determined.pytorch.deepspeed as det_deepspeed
+from determined import workload
+from tests.experiment import utils  # noqa: I100
+from tests.experiment.fixtures import deepspeed_linear_model
+
+deepspeed_config = {
+    "train_batch_size": 16,
+    "train_micro_batch_size_per_gpu": 4,
+    "optimizer": {"type": "SGD", "params": {"lr": 0.001, "weight_decay": 3e-7}},
+    "scheduler": {
+        "type": "WarmupLR",
+        "params": {"warmup_min_lr": 0, "warmup_max_lr": 0.001, "warmup_num_steps": 1000},
+    },
+    "prescale_gradients": False,
+    "fp16": {
+        "enabled": False,
+    },
+    "zero_optimization": {
+        "stage": 0,
+    },
+}
+
+# These environment variables are usually set by the launcher but we set them manually here
+# since they are required by deepspeed.
+os.environ["RANK"] = "0"
+os.environ["LOCAL_RANK"] = "0"
+os.environ["WORLD_SIZE"] = "1"
+os.environ["MASTER_ADDR"] = "localhost"
+os.environ["MASTER_PORT"] = "29500"
+
+
+@pytest.mark.skipif(not torch.cuda.is_available(), reason="no gpu available")
+@pytest.mark.gpu
+class TestDeepSpeedTrial:
+    def setup_method(self) -> None:
+        # This training setup is not guaranteed to converge in general,
+        # but has been tested with this random seed.  If changing this
+        # random seed, verify the initial conditions converge.
+        self.trial_seed = 17
+        self.hparams = {
+            "global_batch_size": 16,
+            "deepspeed_config": deepspeed_config,
+            "test_manual_init_distributed": False,
+            "test_fail_manual_init_distributed": False,
+            "test_manual_dataloader": False,
+            "test_fail_dataset_repro_check": False,
+            "test_manual_grad_acc": False,
+            "test_fail_manual_grad_acc": False,
+            "return_non_scalar_metrics": False,
+            "test_custom_reducer": False,
+        }
+        self.data_parallel_only_auto_train_batch_calls = (
+            deepspeed_config["train_batch_size"]
+            // deepspeed_config["train_micro_batch_size_per_gpu"]
+        )
+
+    def test_fail_manual_init_distributed(self):
+        updated_hparams = copy.deepcopy(self.hparams)
+        updated_hparams["test_fail_manual_init_distributed"] = True
+        os.environ["MANUAL_INIT_DISTRIBUTED"] = "1"
```

Comment:
Let's guarantee that the test isn't leaking environment variables and causing cascading failures.  How about:
```python
os.environ[...] = "1"
try:
    ...
finally:
    del os.environ[...]
```

Also in the tests below.